eyeshade/workers/reports.js

```diff
@@ -815,6 +816,153 @@ const date2objectId = (iso8601, ceilP) => {
                        (ceilP ? 'ffffffffffffffff' : '0000000000000000'))
 }
 
+/**
+ * A referral statement consists of entries describing accrued earnings from referrals,
+ * past successful settlements (payouts), and a balance entry showing a publishers current
+ * settlement balance from referrals.
+ **/
+const referralStatement = async (debug, runtime, owner, summaryP) => {
+  const referrals = runtime.database.get('referrals', debug)
+  const settlements = runtime.database.get('settlements', debug)
+
+  if (!summaryP) {
+    throw new Error('non summary not currently supported')
+  }
+
+  const statements = []
+  const statementTemplate = {
+    referrals: { summary: {}, entries: {} },
+    settlements: { summaries: [], entries: {} },
+    balance: {}
+  }
+
+  const referralFilter = { probi: { $gt: 0 }, altcurrency: { $eq: altcurrency }, exclude: false }
+  const settlementFilter = { probi: { $gt: 0 }, altcurrency: { $eq: altcurrency }, type: { $eq: 'referral' } }
+  if (owner) {
+    referralFilter.owner = { $eq: owner }
+    settlementFilter.owner = { $eq: owner }
+  }
+
+  const referralTotals = await referrals.aggregate([
+    { $match: referralFilter },
+    {
+      $group: {
+        _id: '$publisher',
+        count: { $sum: 1 },
+        probi: { $sum: '$probi' }
+      }
+    }
+  ])
+  referralTotals.forEach((total) => {
+    total.publisher = total._id
+    total.probi = new BigNumber(total.probi)
+
+    if (!statements[total.publisher]) statements[total.publisher] = statementTemplate
+    statements[total.publisher].referrals.summary = total
+  })
+
+  const referralSettlementSummaries = await settlements.aggregate([
+    { $match: settlementFilter },
+    {
+      $group: {
+        // FIXME handle domain transfer case, multiple entries with different owner for same publisher
+        _id: { publisher: '$publisher', currency: '$currency' },
+        amount: { $sum: '$amount' },
+        probi: { $sum: '$probi' },
+        fees: { $sum: '$fees' },
+        commission: { $sum: '$commission' }
+      }
+    }
+  ])
+  referralSettlementSummaries.forEach((summary) => {
+    summary.publisher = summary._id.publisher
+    summary.currency = summary._id.currency
+    summary.probi = new BigNumber(summary.probi)
+
+    if (!statements[summary.publisher]) statements[summary.publisher] = statementTemplate
+    statements[summary.publisher].settlements.summaries.push(summary)
+  })
+
+  underscore.keys(statements).forEach((publisher) => {
+    let pubOwner = ''
+    let balance = new BigNumber(0)
+    if (statements[publisher].referrals.summary.probi) {
+      balance = balance.plus(statements[publisher].referrals.summary.probi)
+      pubOwner = statements[publisher]
+    }
+    statements[publisher].settlements.summaries.forEach((summary) => {
+      balance = balance.minus(summary.probi)
+    })
+    if (balance.lessThan(0)) {
+      throw new Error('Publisher has been overpaid')
```

Comment:
if we're going to throw an error, at the very least, we should indicate *which* publisher; however, i wouldn't throw an error, i'd simply set it to zero and leave it at that. there may be legitimate reasons why past settlements exceeds contributions for a time. it is, of course, undesirable, but it's not clear to me that it's fatal. we could, of course, include a notification to the `eyeshade-bot` if the balance goes negative, so as to alert someone to see what's what... and, of course, please include the publisher-identity that is so afflicated.